internal/docstore/firedocstore/codec.go

```diff
@@ -0,0 +1,271 @@
+// Copyright 2019 The Go Cloud Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package firedocstore
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+
+	"github.com/golang/protobuf/ptypes"
+	"gocloud.dev/internal/docstore/driver"
+	pb "google.golang.org/genproto/googleapis/firestore/v1"
+)
+
+// encodeDoc encodes a driver.Document into Firestore's representation.
+// The Firestore proto definition for Value is a oneof of various types,
+// including basic types like string as well as lists and maps.
+// A Firestore document is just a Go map from strings to Values.
+func encodeDoc(doc driver.Document) (map[string]*pb.Value, error) {
+	var e encoder
+	if err := doc.Encode(&e); err != nil {
+		return nil, err
+	}
+	return e.pv.GetMapValue().Fields, nil
+}
+
+// TODO(jba): support encoding and decoding time.Time and latlng.LatLng specially, since Firestore has
+// special cases for those.
+
+func encodeValue(x interface{}) (*pb.Value, error) {
+	var e encoder
+	if err := driver.Encode(reflect.ValueOf(x), &e); err != nil {
+		return nil, err
+	}
+	return e.pv, nil
+}
+
+// encoder implements driver.Encoder. Its job is to encode a single Firestore value.
+type encoder struct {
+	pv *pb.Value
+}
+
+var nullValue = &pb.Value{ValueType: &pb.Value_NullValue{}}
+
+func (e *encoder) EncodeNil()            { e.pv = nullValue }
+func (e *encoder) EncodeBool(x bool)     { e.pv = &pb.Value{ValueType: &pb.Value_BooleanValue{x}} }
+func (e *encoder) EncodeInt(x int64)     { e.pv = &pb.Value{ValueType: &pb.Value_IntegerValue{x}} }
+func (e *encoder) EncodeUint(x uint64)   { e.pv = &pb.Value{ValueType: &pb.Value_IntegerValue{int64(x)}} }
+func (e *encoder) EncodeBytes(x []byte)  { e.pv = &pb.Value{ValueType: &pb.Value_BytesValue{x}} }
+func (e *encoder) EncodeFloat(x float64) { e.pv = floatval(x) }
+func (e *encoder) EncodeString(x string) { e.pv = &pb.Value{ValueType: &pb.Value_StringValue{x}} }
+
+func (e *encoder) ListIndex(int) { panic("impossible") }
+func (e *encoder) MapKey(string) { panic("impossible") }
+
+func (e *encoder) EncodeComplex(x complex128) {
```

Comment:
I think it will be good to have a list somewhere, probably in the package docstring, explaining:

* Which firestore types are not supported (and what happens if they are encountered)
* Which docstore types are not natively supported by firestore (and how they are represented).

Complex types appear to be 2) from that list, right?

Anyway, doesn't have to be in this PR, but maybe it's easier to generate the list as you go along.

Reply:
Good idea. See comment at top of file.